torchprime/metrics/mfu.py

```diff
@@ -0,0 +1,268 @@
+
+
+@dataclass
+class Config:
+  per_device_batch_size: int
+  max_target_length: int
+  mlp_dim: int
+  emb_dim: int
+  mlp_activations: list[Any]
+  num_experts: int
+  num_experts_per_tok: int
+  num_query_heads: int
+  num_kv_heads: int
+  head_dim: int
+  num_decoder_layers: int
+  vocab_size: int
+  gradient_accumulation_steps: int
+
+
+# Taken from https://github.com/google/maxtext/blob/bd50865c7c66bcf7d12870b8919e59e2a8ebb906/MaxText/maxtext_utils.py#L123
+def calculate_tflops_training_per_device(config: Config, log=True):
+  """Calculate training TFLOP"""
+  ffn1_flops = (
+    2
+    * config.per_device_batch_size
+    * config.max_target_length
+    * config.mlp_dim
+    * config.emb_dim
+    * len(config.mlp_activations)
+  )
+  ffn2_flops = (
+    2
+    * config.per_device_batch_size
+    * config.max_target_length
+    * config.mlp_dim
+    * config.emb_dim
+  )
+  total_ffn_flops = ffn1_flops + ffn2_flops
+
+  if config.num_experts > 1:
+    # MoE: brute force implementation
+    gate_flops = (
+      2
+      * config.per_device_batch_size
+      * config.max_target_length
+      * config.emb_dim
+      * config.num_experts
+    )
+    total_ffn_flops = gate_flops + config.num_experts_per_tok * total_ffn_flops
+
+  qkv_flops = (
+    2
+    * config.per_device_batch_size
+    * config.max_target_length
+    * config.emb_dim
+    * (config.num_query_heads + 2 * config.num_kv_heads)
+    * config.head_dim
+  )
+  attention_flops = (
+    4
+    * config.per_device_batch_size
+    * config.max_target_length**2
+    * config.num_query_heads
+    * config.head_dim
+  )
+  projection_flops = (
+    2
+    * config.per_device_batch_size
+    * config.max_target_length
+    * config.emb_dim
+    * config.num_query_heads
+    * config.head_dim
+  )
+  embedding_flops = (
+    2
+    * config.per_device_batch_size
+    * config.max_target_length
+    * config.emb_dim
+    * config.vocab_size
+  )
+
+  # multiply by 3 for both feed forward and back propagation flops
+  learnable_weight_tflops = (
+    (
+      (total_ffn_flops + qkv_flops + projection_flops) * config.num_decoder_layers
+      + embedding_flops
+    )
+    * 3
+    / 10**12
+  )
+  # megatron tflops calculation does not account for causality in attention
+  attention_tflops = attention_flops * config.num_decoder_layers * 3 / 10**12
+
+  learnable_weight_tflops = learnable_weight_tflops * config.gradient_accumulation_steps
+  attention_tflops = attention_tflops * config.gradient_accumulation_steps
+  total_tflops = learnable_weight_tflops + attention_tflops
+
+  if log:
+    print(
+      "Per train step:\n",
+      f"Total TFLOPs: {total_tflops:.2f} \n",
+      f"split as {100 * learnable_weight_tflops/total_tflops:.2f}% learnable weight flops",
+      f"and {100 * attention_tflops/total_tflops:.2f}% attention flops",
+    )
+  return total_tflops, learnable_weight_tflops, attention_tflops
```

Comment:
nit: Should we just return total_tflops? or Are the other things returned just for testing?

Reply:
Removed. MaxText did this and I just copied it :)